File: android/core/network/src/main/kotlin/dev/soupy/eclipse/android/core/network/AniSkipService.kt
package dev.soupy.eclipse.android.core.network

import dev.soupy.eclipse.android.core.model.SkipSegment
import dev.soupy.eclipse.android.core.model.SkipType
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.decodeFromString

class AniSkipService(
    private val baseUrl: String = "https://api.aniskip.com/v2",
    private val httpClient: EclipseHttpClient = EclipseHttpClient(),
) {
    suspend fun fetchSkipTimes(
        malId: Int,
        episodeNumber: Int,
        episodeDurationSeconds: Double,
    ): NetworkResult<List<SkipSegment>> {
        val url = aniSkipTimesUrl(
            baseUrl = baseUrl,
            malId = malId,
            episodeNumber = episodeNumber,
            episodeDurationSeconds = episodeDurationSeconds,
        )

        return when (val result = httpClient.get(url)) {
            is NetworkResult.Success -> decode(result.value, episodeDurationSeconds)
            is NetworkResult.Failure.Http -> NetworkResult.Success(emptyList())
            is NetworkResult.Failure.Connectivity -> result
            is NetworkResult.Failure.Serialization -> result
        }
    }

    private fun decode(
        body: String,
        episodeDurationSeconds: Double,
    ): NetworkResult<List<SkipSegment>> = try {
        val response = EclipseJson.decodeFromString<AniSkipResponse>(body)
        val segments = if (response.found) {
            response.results.orEmpty()
                .mapNotNull { result ->
                    val type = result.skipType.toSkipType() ?: return@mapNotNull null
                    SkipSegment(
                        startTime = result.interval.startTime,
                        endTime = result.interval.endTime,
                        type = type,
                    ).clamped(episodeDurationSeconds)
                }
                .sortedBy(SkipSegment::startTime)
        } else {
            emptyList()
        }

        NetworkResult.Success(segments)
    } catch (error: SerializationException) {
        NetworkResult.Failure.Serialization(error)
    }
}

internal fun aniSkipTimesUrl(
    baseUrl: String,
    malId: Int,
    episodeNumber: Int,
    episodeDurationSeconds: Double,
): String {
    val episodeLength = if (episodeDurationSeconds.isFinite() && episodeDurationSeconds > 0) {
        episodeDurationSeconds.toInt()
    } else {
        0
    }
    return "$baseUrl/skip-times/$malId/$episodeNumber" +
        "?types%5B%5D=op&types%5B%5D=ed&types%5B%5D=recap" +
        "&types%5B%5D=mixed-op&types%5B%5D=mixed-ed&episodeLength=$episodeLength"
}

@Serializable
private data class AniSkipResponse(
    val found: Boolean,
    val results: List<AniSkipResult>? = null,
    val statusCode: Int = 0,
)

@Serializable
private data class AniSkipResult(
    val interval: AniSkipInterval,
    val skipType: String,
    val skipId: String? = null,
    val episodeLength: Double = 0.0,
)

@Serializable
private data class AniSkipInterval(
    val startTime: Double,
    val endTime: Double,
)

private fun String.toSkipType(): SkipType? = when (this) {
    "op", "mixed-op" -> SkipType.INTRO
    "ed", "mixed-ed" -> SkipType.OUTRO
    "recap" -> SkipType.RECAP
    else -> null
}

File: Eclipse/Player/AniSkipService.swift
//
//  AniSkipService.swift
//  Eclipse
//
//  Created on 27/02/26.
//

import Foundation

// MARK: - Skip Segment Models

enum SkipType: String, Codable {
    case intro
    case outro
    case recap
    case preview

    var displayLabel: String {
        switch self {
        case .intro: return "Skip Intro"
        case .outro: return "Skip Outro"
        case .recap: return "Skip Recap"
        case .preview: return "Skip Preview"
        }
    }
}

struct SkipSegment {
    let startTime: Double
    let endTime: Double
    let type: SkipType

    /// Unique key used to track whether this segment has already been auto-skipped.
    var uniqueKey: String {
        guard startTime.isFinite, startTime >= 0, startTime <= Double(Int.max) else {
            return "\(type.rawValue)_unknown"
        }
        return "\(type.rawValue)_\(Int(startTime))"
    }
}

// MARK: - AniSkip API Response Models

private struct AniSkipResponse: Codable {
    let found: Bool
    let results: [AniSkipResult]?
    let statusCode: Int
}

private struct AniSkipResult: Codable {
    let interval: AniSkipInterval
    let skipType: String
    let skipId: String
    let episodeLength: Double
}

private struct AniSkipInterval: Codable {
    let startTime: Double
    let endTime: Double
}

// MARK: - AniSkip Service

final class AniSkipService {
    static let shared = AniSkipService()

    private let baseURL = "https://api.aniskip.com/v2"
    private let session: URLSession

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 15
        self.session = URLSession(configuration: config)
    }

    /// Fetches skip-time segments for an anime episode using MyAnimeList ID.
    /// - Parameters:
    ///   - malId: The MyAnimeList media ID for the specific season / anime.
    ///   - episodeNumber: The episode number within that season.
    ///   - episodeDuration: The total duration in seconds (used for better matching).
    /// - Returns: Array of skip segments (intro, outro, recap).
    func fetchSkipTimes(malId: Int, episodeNumber: Int, episodeDuration: Double) async throws -> [SkipSegment] {
        let durationIsUsable = episodeDuration.isFinite && episodeDuration > 0
        let episodeLength = durationIsUsable ? Int(episodeDuration) : 0
        let durationParam = "&episodeLength=\(episodeLength)"
        let urlString = "\(baseURL)/skip-times/\(malId)/\(episodeNumber)?types[]=op&types[]=ed&types[]=recap&types[]=mixed-op&types[]=mixed-ed\(durationParam)"

        guard let url = URL(string: urlString) else {
            Logger.shared.log("AniSkipService: Invalid URL: \(urlString)", type: "Error")
            return []
        }

        Logger.shared.log("AniSkipService: Fetching skip times for malId=\(malId) ep=\(episodeNumber) duration=\(formatSeconds(episodeDuration))", type: "AniSkip")

        let (data, response) = try await session.data(from: url)

        guard let httpResponse = response as? HTTPURLResponse else {
            Logger.shared.log("AniSkipService: Non-HTTP response", type: "Error")
            return []
        }

        guard httpResponse.statusCode == 200 else {
            Logger.shared.log("AniSkipService: HTTP \(httpResponse.statusCode) for malId=\(malId) ep=\(episodeNumber)", type: "AniSkip")
            return []
        }

        let decoded = try JSONDecoder().decode(AniSkipResponse.self, from: data)

        guard decoded.found, let results = decoded.results else {
            Logger.shared.log("AniSkipService: No skip times found for malId=\(malId) ep=\(episodeNumber)", type: "AniSkip")
            return []
        }

        let segments = results.compactMap { result -> SkipSegment? in
            let skipType: SkipType
            switch result.skipType {
            case "op", "mixed-op":
                skipType = .intro
            case "ed", "mixed-ed":
                skipType = .outro
            case "recap":
                skipType = .recap
            default:
                return nil
            }

            guard result.interval.startTime.isFinite, result.interval.endTime.isFinite else { return nil }
            let start = max(0, result.interval.startTime)
            let end = durationIsUsable ? min(episodeDuration, result.interval.endTime) : result.interval.endTime
            guard end > start else { return nil }

            return SkipSegment(startTime: start, endTime: end, type: skipType)
        }

        Logger.shared.log(
            "AniSkipService: Found \(segments.count) skip segments for malId=\(malId) ep=\(episodeNumber): "
            + segments.map { "\($0.type.rawValue) \(formatSeconds($0.startTime))-\(formatSeconds($0.endTime))s" }.joined(separator: ", "),
            type: "AniSkip"
        )

        return segments
    }

    private func formatSeconds(_ value: Double) -> String {
        guard value.isFinite else { return "nil" }
        return "\(Int(value.rounded()))"
    }
}
